Allow validating unicode signatures.

ralphbean opened the pull request:
I just noticed this in the logs of program driven by `fedmsg.tail_messages`.
The signature was unicode, and so this type check failed every time.

This seems to have been inadvertently broken in f0553d285e586dc48f1a244b81f1517551d469ee ?

ralphbean committed:
Allow validating unicode signatures.

I just noticed this in the logs of program driven by `fedmsg.tail_messages`.
The signature was unicode, and so this type check failed every time.

This seems to have been inadvertently broken in f0553d285e586dc48f1a244b81f1517551d469ee ?

jeremycline commented on fedmsg/crypto/x509.py:
```diff
@@ -113,7 +113,7 @@ def fail(reason):
         if field not in message:
             return fail("No %r field found." % field)
-        if not isinstance(message[field], str):
+        if not isinstance(message[field], basestring):
```
``basestring`` is undefined in Python 3. While this code will never actually run in Python 3 since it's the M2Crypto implementation, this makes the API unclear and it should either be ``six.text_type`` (unicode and str in Python 2 and Python 3, respectively) or ``six.binary_type`` (str and bytes in Python 2 and Python 3, respectively).

Since we don't know the encoding at this point, I believe it should be ``six.text_type``.

ralphbean replied:
That's legit.

The underlying issue is (I think) that some routines pass unicode signatures in here, while others pass byte-strings.

I guess I'm with you.. that we should enforce "unicode" as the api with ``six.text_type`` and then go looking for any places that may erroneously call this with byte strings.

jeremycline reviewed (changes requested):
Thanks for the PR! 

This change needs tests to demonstrate the issue before it can be merged.

jeremycline committed:
Assert the signature/certificate fields are unicode

This checks the certificate and signature fields in messages to ensure
they're text. If they're not, an error is logged and we do our best to
decode the message as UTF-8. This decoding is a short-term measure until
the users of the APIs have been hunted down and fixed.

Signed-off-by: Jeremy Cline <jeremy@jcline.org>

jeremycline committed:
Move fedmsg.crypto.x509 ``log`` to ``_log``

This is mostly because it makes writing tests easier, and also because
it shouldn't be a public symbol.

Signed-off-by: Jeremy Cline <jeremy@jcline.org>

jeremycline committed:
Tests for PR 456

Signed-off-by: Jeremy Cline <jeremy@jcline.org>

jeremycline committed:
Add two tests asserting the correct sign API

``fedmsg.crypto.sign`` should return unicode text signatures and
certificates, but it currently does not. These tests assert the correct
API and are marked as expected failures.

Signed-off-by: Jeremy Cline <jeremy@jcline.org>

jeremycline committed:
Add the encoding header to test_x509.py

There are non-ascii characters in the file so it needs an encoding
header.

Signed-off-by: Jeremy Cline <jeremy@jcline.org>

jeremycline committed:
Remove some unreachable code

The dictionary is checked for both keys above and thus this block cannot
be reached.

Signed-off-by: Jeremy Cline <jeremy@jcline.org>

jeremycline committed:
Add a test to assert the validate API accepts text signatures

Signed-off-by: Jeremy Cline <jeremy@jcline.org>

puiterwijk commented:
Quick glance says that this looks okay to me.

ralphbean commented:
> Quick glance says that this looks okay to me.

Same.

jeremycline reviewed (approved):
:shipit:

jeremycline commented:
This patch was backported to 0.19.1
